jetty-maven-plugin/src/main/java/org/eclipse/jetty/maven/plugin/JettyStopMojo.java

```diff
@@ -72,38 +74,132 @@ public void execute() throws MojoExecutionException, MojoFailureException
-            out.write((stopKey + "\r\n" + command + "\r\n").getBytes());
-            out.flush();
-
-            if (stopWait > 0)
+            //try to get the pid of the forked jetty process
+            Long pid = null;
+            try
+            {
+                String response = send(stopKey + "\r\n" + "pid" + "\r\n", true, stopWait);
+                pid = Long.valueOf(response);
+            }
+            catch (NumberFormatException e)
             {
-                s.setSoTimeout(stopWait * 1000);
-                s.getInputStream();
+                getLog().info("Server returned bad pid");
+            }
+            catch (ConnectException e)
+            {
+                //jetty not running, no point continuing
+                getLog().info("Jetty not running!");
+                return;
+            }
+            catch (Exception e)
+            {
+                //jetty running, try to stop it regardless of error
+                getLog().error(e);
+            }
 
-                getLog().info("Waiting " + stopWait + " seconds for jetty to stop");
-                LineNumberReader lin = new LineNumberReader(new InputStreamReader(s.getInputStream()));
-                String response;
-                boolean stopped = false;
-                while (!stopped && ((response = lin.readLine()) != null))
+            //now send the stop command and wait for confirmation - either an ack from jetty, or
+            //that the process has stopped
+            try
+            {
+                if (pid == null)
                 {
+                    //no pid, so just wait until jetty reports itself stopped
+                    getLog().info("Waiting " + stopWait + " seconds for jetty to stop");
+                    String response = send(stopKey + "\r\n" + command + "\r\n", true, stopWait);
+
                     if ("Stopped".equals(response))
-                    {
-                        stopped = true;
                         getLog().info("Server reports itself as stopped");
+                    else
+                        getLog().info("Couldn't verify server as stopped, received " + response);
+                }
+                else
+                {
+                    //wait for pid to stop
+                    getLog().info("Waiting " + stopWait + " seconds for jetty " + pid + " to stop");
+                    send(stopKey + "\r\n" + command + "\r\n", false, 0);
+                    Optional<ProcessHandle> optional = ProcessHandle.of(pid);
+                    ProcessHandle handle = optional.orElse(null);
+                    if (handle != null)
+                    {
+                        try
+                        {
+                            CompletableFuture<ProcessHandle> future = handle.onExit();
+                            if (handle.isAlive())
+                            {
+                                handle = future.get(stopWait, TimeUnit.SECONDS);
+                            }
```

Comment:
Don't think this block is correct.
`future.get(wait, unit)` throws `TimeoutException` if the time elapses, but the code seems to hint something like "if timeout and still alive, do something", so the "do something" will be skipped.

Have you considered using `ProcessHandle.destroy[Forcibly]()` if the timeout expires?
Perhaps along with `CompletableFuture.orTimeout()`?

Reply:
@sbordet this would be a behaviour change of the JettyStopMojo:  it has always just made a best effort to stop the remote jetty, waiting up to the configured time to get an acknowledgement from jetty that it is shutting down before giving up. I suppose we could add a new configuration option to forcibly destroy the process after the timeout expires.